htlcswitch/link.go

```diff
@@ -639,20 +653,49 @@ func (l *channelLink) WaitForShutdown() {
 // actively accept requests to forward HTLC's. We're able to forward HTLC's if
 // we are eligible to update AND the channel isn't currently flushing the
 // outgoing half of the channel.
+//
+// NOTE: MUST NOT be called from the main event loop.
 func (l *channelLink) EligibleToForward() bool {
-	return l.EligibleToUpdate() &&
-		!l.IsFlushing(Outgoing)
+	return runInEventLoop(l, fn.Unit{}, func(_ fn.Unit) bool {
+		return l.eligibleToForward()
+	}).UnwrapLeftOr(false)
+}
+
+// eligibleToForward returns a bool indicating if the channel is able to
+// actively accept requests to forward HTLC's. We're able to forward HTLC's if
+// we are eligible to update AND the channel isn't currently flushing the
+// outgoing half of the channel.
+//
+// NOTE: MUST be called from the main event loop.
+func (l *channelLink) eligibleToForward() bool {
+	return l.eligibleToUpdate() && !l.IsFlushing(Outgoing)
 }
 
 // EligibleToUpdate returns a bool indicating if the channel is able to update
 // channel state. We're able to update channel state if we know the remote
 // party's next revocation point. Otherwise, we can't initiate new channel
 // state. We also require that the short channel ID not be the all-zero source
 // ID, meaning that the channel has had its ID finalized.
+//
+// NOTE: MUST be called from the main event loop.
 func (l *channelLink) EligibleToUpdate() bool {
+	return runInEventLoop(l, fn.Unit{}, func(_ fn.Unit) bool {
+		return l.eligibleToUpdate()
```

Comment:
Is this the agreed upon philosophy? I generally don't like shipping incomplete APIs just because we don't make use of them. It leads to a lot more inconsistent style when the APIs get added later. Also it means that when someone needs this later it's already available as opposed to them having to reverse engineer how to do it on a yak shaving quest to do what they wanted to do.

Reply:
> Is this the agreed upon philosophy?

Personal credo: less code = moar good, can't speak to the lnd-project-as-a-whole philosophy so I defer to @yyforyongyu for second opinion 🫡

Reply:
this is minor IMO, but agree with @carlaKC. Don't wanna throw a full paragraph here but in general either "add minimal APIs => use it => add more => refactor" or "bring up a good design => less refactor". For switch code I'd say the first works in reality.